File: clients/typescript/src/action_client.ts
import { ulid } from 'ulid';
import { ActionRequest } from './dtos';
import { resourceManager } from './resource_manager';
import { clientSetup, postJson, AmenoException } from './shared';

export class Action {
    public readonly name: string;

    constructor(name: string) {
        this.name = name;
    }

    public async request<T>(payload?: unknown): Promise<T> {
        const requestDto: ActionRequest<unknown> = {
            id: ulid(),
            createdAt: new Date().toISOString(),
            type: 'ActionRequest',
            route: this.name,
            payload,
            appName: clientSetup.appName
        };

        const url = `${clientSetup.originUrl}/api/request`;
        const responseData = await postJson<any>(url, requestDto);

        if (responseData?.success !== true) {
            const errorInfo = responseData?.error;
            let errorMessage = 'Erro desconhecido ao executar ação.';

            if (typeof errorInfo === 'object' && errorInfo !== null && errorInfo.message)
                errorMessage = errorInfo.message;
            else if (typeof responseData?.errorMessage === 'string' && responseData.errorMessage)
                errorMessage = responseData.errorMessage;

            throw new AmenoException(errorMessage);
        }

        const responseValue = 'result' in responseData ? responseData.result : responseData.response;
        return responseValue as T;
    }

    public async queue(payload?: unknown): Promise<void> {
        const requestDto: ActionRequest<unknown> = {
            id: ulid(),
            createdAt: new Date().toISOString(),
            type: 'ActionRequest',
            route: this.name,
            payload,
            appName: clientSetup.appName
        };

        const url = `${clientSetup.originUrl}/api/queue`;
        await postJson(url, requestDto);
    }
}

export function action(name: string): Action {
    resourceManager.actions.add(name);
    return new Action(name);
}

File: clients/csharp/AmenoLink/ActionClient.cs
using System.Text.Json;

namespace AmenoLink;

public class ActionClient(string name) : IAction
{
    public string Name { get; } = name;

    public async Task<T> Request<T>(object? payload = null)
    {
        var requestDto = new ActionRequest<object?>(
            Id: Ulid.NewUlid().ToString(),
            CreatedAt: DateTime.UtcNow.ToString("o"),
            Route: Name,
            Type: "ActionRequest",
            Payload: payload,
            AppName: AmenoLinkClient.Settings.AppName
        );

        string url = $"{AmenoLinkClient.Settings.OriginUrl}/api/request";
        var responseElement = await AmenoLinkClient.PostJson<JsonElement>(url, requestDto);

        bool success = responseElement.TryGetProperty("success", out var successProperty) && successProperty.GetBoolean();
        if (!success)
        {
            string errorMessage = "Erro desconhecido ao executar ação.";

            if (responseElement.TryGetProperty("error", out var errorProperty) && errorProperty.ValueKind == JsonValueKind.Object)
            {
                if (errorProperty.TryGetProperty("message", out var messageProperty))
                    errorMessage = messageProperty.GetString() ?? errorMessage;
            }
            else if (responseElement.TryGetProperty("errorMessage", out var errorMessageProperty))
                errorMessage = errorMessageProperty.GetString() ?? errorMessage;

            throw new AmenoException(errorMessage);
        }

        if (responseElement.TryGetProperty("result", out var resultElement))
            return JsonSerializer.Deserialize<T>(resultElement.GetRawText(), AmenoLinkClient.JsonOptions)!;

        if (responseElement.TryGetProperty("response", out var responsePropertyLegacy))
            return JsonSerializer.Deserialize<T>(responsePropertyLegacy.GetRawText(), AmenoLinkClient.JsonOptions)!;

        return default!;
    }

    public async Task Queue(object? payload = null)
    {
        var requestDto = new ActionRequest<object?>(
            Id: Ulid.NewUlid().ToString(),
            CreatedAt: DateTime.UtcNow.ToString("o"),
            Route: Name,
            Type: "ActionRequest",
            Payload: payload,
            AppName: AmenoLinkClient.Settings.AppName
        );

        string url = $"{AmenoLinkClient.Settings.OriginUrl}/api/queue";
        await AmenoLinkClient.PostJson<object?>(url, requestDto);
    }
}
